txhelpers/cspp.go

```diff
@@ -0,0 +1,143 @@
+package txhelpers
+
+import (
+	"github.com/decred/dcrd/dcrutil/v2"
+	"github.com/decred/dcrd/wire"
+)
+
+// must be sorted large to small
+var splitPoints = [...]dcrutil.Amount{
+	1 << 36, // 687.19476736
+	1 << 34, // 171.79869184
+	1 << 32, // 042.94967296
+	1 << 30, // 010.73741824
+	1 << 28, // 002.68435456
+	1 << 26, // 000.67108864
+	1 << 24, // 000.16777216
+	1 << 22, // 000.04194304
+	1 << 20, // 000.01048576
+	1 << 18, // 000.00262144
+}
+
+var splitPointMap = map[int64]struct{}{}
+
+func init() {
+	for _, amt := range splitPoints {
+		splitPointMap[int64(amt)] = struct{}{}
+	}
+}
+
+// IsMixTx test if a transaction is a CSPP-mixed transaction, which must have 3
+// or more outputs of the same amount, which is one of the pre-defined mix
+// denominations. mixDenom is the largest of such denominations. mixCount is the
+// number of inputs of this denomination.
+func IsMixTx(tx *wire.MsgTx) (isMix bool, mixDenom int64, mixCount uint32) {
+	if len(tx.TxOut) < 3 {
+		return false, 0, 0
+	}
+
+	mixedOuts := make(map[int64]uint32)
+	for _, o := range tx.TxOut {
+		val := o.Value
+		if _, ok := splitPointMap[val]; ok {
+			mixedOuts[val]++
+			continue
+		}
+	}
+
+	for val, count := range mixedOuts {
+		if count < 3 {
+			continue
+		}
+		if val > mixDenom {
+			mixDenom = val
+			mixCount = count
+		}
+	}
+
+	isMix = mixCount >= uint32(len(tx.TxOut)/2)
+	return
+}
+
+// FeeForSerializeSize calculates the required fee for a transaction of some
+// arbitrary size given a mempool's relay fee policy.
+func FeeForSerializeSize(relayFeePerKb dcrutil.Amount, txSerializeSize int) dcrutil.Amount {
+	fee := relayFeePerKb * dcrutil.Amount(txSerializeSize) / 1000
+
+	if fee == 0 && relayFeePerKb > 0 {
+		fee = relayFeePerKb
+	}
+
+	if fee < 0 || fee > dcrutil.MaxAmount {
+		fee = dcrutil.MaxAmount
+	}
+
+	return fee
+}
+
+// DefaultRelayFeePerKb is the default minimum relay fee policy for a mempool.
+const DefaultRelayFeePerKb int64 = 1e4
+
+// The size of a solo (non-pool) ticket purchase transaction assumes a specific
+// transaction structure and worst-case signature script sizes.
+func calcSoloTicketTxSize() int {
+	inSizes := []int{RedeemP2PKHSigScriptSize}
+	outSizes := []int{P2PKHPkScriptSize + 1, TicketCommitmentScriptSize, P2PKHPkScriptSize + 1}
+	return EstimateSerializeSizeFromScriptSizes(inSizes, outSizes, 0)
+}
+
+var (
+	soloTicketTxSize    = calcSoloTicketTxSize()
+	defaultFeeForTicket = FeeForSerializeSize(dcrutil.Amount(DefaultRelayFeePerKb), soloTicketTxSize)
```

Comment:
This is another thing that only allows `IsMixedSplitTx` to detect mixing ticket split transactions created with the *current* CSPP server, which computes it's output amount for such transactions in this way.